corporate/tests/test_stripe.py

```diff
@@ -5432,3 +5435,82 @@ def test_downgrade_realm_and_void_open_invoices(self, *mocks: Mock) -> None:
+        hamlet = self.example_user("hamlet")
+        realm = hamlet.realm
+
+        self_hosted_billing_url = "/self-hosted-billing/"
+        result = self.client_get(self_hosted_billing_url)
+        self.assertEqual(result.status_code, 302)
+        self.assertIn("http://selfhosting.testserver/remote-billing-login/", result["Location"])
+
+        signed_auth_url = result["Location"]
+        signed_access_token = signed_auth_url.split("/")[-1]
+        result = self.client_get(signed_auth_url, subdomain="selfhosting")
+        # When logging in for the first time some extra steps are needed
+        # to confirm and verify the email address.
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_success_response(["Enter email"], result)
+        self.assert_in_success_response(
+            [f'action="/remote-billing-login/{signed_access_token}/confirm/"'], result
+        )
+
+        result = self.client_post(
+            f"/remote-billing-login/{signed_access_token}/confirm/",
+            {"email": hamlet.delivery_email},
+            subdomain="selfhosting",
+        )
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_success_response(
+            [
+                "To finish logging in, check your email account (",
+                ") for a confirmation email from Zulip.",
+                hamlet.delivery_email,
+            ],
+            result,
+        )
+
+        confirmation_url = self.get_confirmation_url_from_outbox(
+            hamlet.delivery_email,
+            url_pattern=(
+                f"{settings.SELF_HOSTING_MANAGEMENT_SUBDOMAIN}.{settings.EXTERNAL_HOST}"
+                r"(\S+)"
+            ),
+            email_body_contains="confirm your email and log in to Zulip plan management",
+        )
+        result = self.client_get(confirmation_url, subdomain="selfhosting")
+        self.assertEqual(result.status_code, 302)
+        self.assertTrue(result["Location"].startswith("/remote-billing-login/"))
+    
+        result = self.client_get(result["Location"], subdomain="selfhosting")
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_success_response(["Log in to Zulip plan management"], result)
+
+        result = self.client_post(signed_auth_url, {"tos_consent": "true"}, subdomain="selfhosting")
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(result["Location"], f"/realm/{realm.uuid!s}/plans/")
+
+        # upgrade to business plan
+        result = self.client_get(f"/realm/{realm.uuid!s}/upgrade/", subdomain="selfhosting")
+        self.assertEqual(result.status_code, 200)
+        self.assert_in_success_response(["Add card", "Purchase Zulip Business"], result)
+
+        self.assertFalse(Customer.objects.exists())
+        self.assertFalse(CustomerPlan.objects.exists())
+        self.assertFalse(LicenseLedger.objects.exists())
+
+        stripe_customer = self.add_card_and_upgrade(hamlet)
```

Comment:
Pointing out this does work to test the remote realm flow. You need to modify the function to test the upgrade for remote realm and not the realm.